EricXiong420 iP

EricXiong420 opened the pull request:
### Hachi

> "I long to accomplish a great and noble task, but it is my chief duty to accomplish small tasks as if they were great and noble." - Helen Keller ([source](https://www.brainyquote.com/quotes/helen_keller_114884?src=t_tasks))

Hachi frees your mind from the things you have to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/EricXiong420/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Notifications (coming soon)

> [!NOTE]
> You may become addicted to using this task list 🙃.

![an abstract painting of a to-do list](https://i.etsystatic.com/23353241/r/il/48083b/2319753672/il_fullxfull.2319753672_68ka.jpg)

---

If you're a Java programmer, you can use it to practice Java too! Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        new Hachi(DEFAULT_TASK_PATH).run();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

EricXiong420 committed:
Level 0. Renamed Chatbot and wrote skeletal version that prints a hello
and goodbye message

EricXiong420 committed:
Makes the program echo user inputs

EricXiong420 committed:
Allows the building of a list storing data

EricXiong420 committed:
Adds checkboxes to the list that can be marked and unmarked

EricXiong420 committed:
Adds three types of tasks which each have their representation

EricXiong420 committed:
Added semi-automated testing under text-ui-test

EricXiong420 committed:
Added and catches every possible exception in the program

EricXiong420 committed:
Added the delete command

EricXiong420 committed:
Allow saving and loading tasks from hard drive

Tasks are currently saved in an ArrayList, which disappears when the
program restarts.

Saving the tasks such that they are persistent will make the program
more useful and prevent data from being lost when the program quits.

Let's store the tasks in a text file in the data directory and update it
accordingly when the task list changes, and obtain tasks from it when
the program starts.

EricXiong420 committed:
Merge branch 'branch-Level-7'

EricXiong420 committed:
Allow chatbot to understand dates and times

Previously, chatbot saved dates only as strings, which meant the user
could put in anything and it would have been accepted.

Saving dates as actual dates would allow better error checking and more
filtering functionality.

Let's save dates and print it out in a certain format, and add in a
search date function to demonstrate it's utility.

EricXiong420 committed:
Merge branch 'branch-Level-8'

EricXiong420 committed:
Add Ui, Parser, Command, Storage objects

Previously, the program was not as Object-Oriented.

In order to make the code more readable, the code will be split into the
4 files mentioned above, as well as Command objects.

So far, have implemented some methods in Ui, Parser, and Command.

EricXiong420 committed:
Move corresponding code to their objects

Code in Hachi.java was not completely separated out into separate
objects.

Let's move the code to their respective objects, making the code more
object-oriented.

EricXiong420 committed:
Move code to their respective objects

The previous commit only included one file.

Let's commit the rest of the files that will help make the code more
object oriented.

EricXiong420 committed:
Merge branch 'branch-A-MoreOOP'

EricXiong420 committed:
Organise the class into a package

The classes were originally not within a package.

Let's put them all into one package called hachi.

EricXiong420 committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

EricXiong420 committed:
Set up Gradle support

Project was previously not using gradle for managing dependencies,
running tests, checking style etc.

Let's add gradle so that a lot of these tasks can be automated.

EricXiong420 committed:
Add testing with JUnit

Previously did not employ unit testing to ensure that program is bug
free.

Let's now use JUnit to test the components and methods of our program.

EricXiong420 committed:
Add 'find' command to search for tasks

Previously could not search for tasks based on the task name.

Let's allow people to search for tasks by typing in a phrase that could
correspond to a part of the task name.

EricXiong420 committed:
Add Javadoc comments to public classes and methods.

Previously there was insufficient Javadoc comments and documentation.

Let's add Javadoc comments to all the public classes and methods.

EricXiong420 committed:
Follow coding standard

Previously code had many places which did not follow the coding
standard.

A consistent coding standard allows for more readable and maintainable
code.

Let's change those parts which do not follow the coding standard to
follow it.

EricXiong420 committed:
Merge branch 'branch-A-CodingStandard'

EricXiong420 committed:
Merge branch 'branch-A-JavaDoc'

EricXiong420 committed:
Merge branch 'branch-Level-9'

EricXiong420 committed:
Add CheckStyle

This is a dummy commit, as checkstyle was already added beforehand.

In general, checkstyle was added to make it easier to keep track of
coding standard violations throughout the code.

EricXiong420 committed:
Merge branch 'branch-A-CheckStyle'

Jonyxzx commented on src/main/java/hachi/Hachi.java:
```diff
@@ -0,0 +1,231 @@
+     * Runs the task list program.
+     * @throws HachiException
+     */
+    public void run() throws HachiException {
```
The run function has many logical operations with multiple if-else brancing. Perhaps some helper function of higher level abstraction could help the code to be more readable

Jonyxzx commented on src/main/java/hachi/Parser.java:
```diff
@@ -0,0 +1,219 @@
+        return number - 1;
+    }
+
+    private static Task convertTextToTask(String txt) throws HachiException {
```
This function includes multiple try catch and nested if statement, could consider breaking them into multiple functions

Jonyxzx commented on src/main/java/hachi/Parser.java:
```diff
@@ -0,0 +1,219 @@
+     */
+    public static void checkArgumentLength(String commandWord, int argumentLength)
+            throws HachiException {
+        switch (commandWord) {
```
Great use of switch case to handle multiple events

Jonyxzx commented on src/main/java/hachi/Parser.java:
```diff
@@ -0,0 +1,219 @@
+     * @return The final parsed string.
+     */
+    public static String parseTaskArguments(String taskType, String[] arg) {
+        switch (taskType) {
```
would if-else statement be sufficient for a single case?

Jonyxzx commented on src/main/java/hachi/Ui.java:
```diff
@@ -0,0 +1,105 @@
+     * @return true if the user raw input line should be ignored.
+     */
+    private boolean shouldIgnore(String rawInputLine) {
+        return rawInputLine.trim().isEmpty() || isCommentLine(rawInputLine);
```
Great use of abstraction to simplify code reading

Jonyxzx reviewed (commented):
Overall, good effort to handle exceptions and commands. Some functions are too long and hard to read othewise amazing well designed code

EricXiong420 committed:
Add graphical user interface

Previously, the program used a text-based interface.

Let's move to a graphical user interface that will be more pleasing to
the eye and perhaps easier to use.

adammangzijun commented on src/main/java/hachi/Command.java:
```diff
@@ -0,0 +1,22 @@
+    }
+
+    public String getCommand() {
+        return commandWord;
```
Maybe should have added this.?

adammangzijun commented on src/main/java/hachi/Event.java:
```diff
@@ -0,0 +1,59 @@
+    private LocalDate endDate;
+
+    /**
+     * Constructor for the Event class.
```
I think there should be an empty line between the description and the parameters.

adammangzijun commented on src/main/java/hachi/JavaFx.java:
```diff
@@ -0,0 +1,134 @@
+package hachi;
+
+import exceptions.HachiException;
```
Should you separate the imports from different packages with empty lines?

adammangzijun commented on src/main/java/hachi/Parser.java:
```diff
@@ -0,0 +1,219 @@
+            }
+            break;
+        default:
+            break;
```
I was thinking the default case should do something like printing out errors rather than just a break statement.

adammangzijun commented on src/main/java/hachi/Task.java:
```diff
@@ -0,0 +1,63 @@
+
+    /**
+     * Returns true if string is within the task name.
+     *
```
There are inconsistencies in your javaDoc, some have a line between description and params, some do not.

adammangzijun reviewed (commented):
Overall well done in coding style, the most important takeaway should be your javaDocs. 😄

EricXiong420 committed:
Add GUI

EricXiong420 committed:
Merge branch 'branch-Level-10'

EricXiong420 committed:
Change gradle main method to target Launcher

Previously, the main method in gradle was assigned to that in the Hachi
class. As a result, the jar file created would use the main method in
Hachi, which did not have a graphical user interface.

Providing the user a graphical user interface will allow a more
streamlined and interactive user experience.

Let's change the build.gradle file to point to the main method in
Launcher, so the user can now use the GUI in the jar file.

EricXiong420 committed:
Abstract out all the user commands

Previously the user commands were all handled in one giant method.
Abstracting them out and using method overriding to handle all the
commands would make the program much more readable and the code quality
would be much higher.

Let's create a separate Command class for every command, and move each
command's execution body to the command's "execute()" method.

EricXiong420 committed:
Merge pull request #1 from EricXiong420/branch-A-CodeQuality

Abstract out all the user commands

EricXiong420 committed:
Add assertions

EricXiong420 committed:
Merge pull request #2 from EricXiong420/branch-A-Assertions

Add assertions

EricXiong420 committed:
Add ability to sort tasks by name when listing

The user may wish to sort the tasks by alphabetical order.

Let's provide a way to do so under the "list" command.

EricXiong420 committed:
Merge branch 'branch-BCD-Extension'

EricXiong420 committed:
Add comments

EricXiong420 committed:
Change completed boolean name

Previously, the boolean "completed" did not abide by the naming
conventions.

Let's change it to "isCompleted" so that it does.

EricXiong420 committed:
Add user guide

Let's add a user guide so that the user of the program will be able to
know what commands he or she can use.

EricXiong420 committed:
Improve GUI

Let's improve the GUI to be more pleasing to look at by making the
profile pictures circular, and the text bubbles rounded rectangles!

EricXiong420 committed:
Add Ui image
